related-files.el

```diff
@@ -234,35 +247,71 @@ ones.  Look at `related-files-jump' and `related-files-make' if you don't
 want to mix existing and non-existing files in the same list..
 
 If JUMPERS is not provided, use `related-files-jumpers'.  If
-CURRENT-PLACE is not provided, use the function
-`buffer-file-name'.
+CURRENT-PLACES is not provided, use
+`related-files-current-place-finders'.
 
 Interactively, a numeric prefix argument selects the jumper at
 the specified position (zero-based index) in `related-files-jumpers'."
   (interactive (list (when (numberp current-prefix-arg)
                        (list (seq-elt related-files-jumpers current-prefix-arg)))))
-  (related-files--jump-or-make jumpers current-place
+  (related-files--jump-or-make jumpers current-places
                          :include-existing-places t
                          :include-non-existing-places t))
 
 
-;;; Jumpers Public API
+;;; Jumpers and Places Public API
 
 (cl-defgeneric related-files-apply (jumper place)
   "Apply JUMPER to PLACE and return related places or nil.
 
 PLACE is a filename and the result must be a possibly-empty list
-of filenames.
+of filenames.")
 
-The default implementation allows JUMPER to be a function.  The
-function can return either a single place or a possibly-empty
-list of places."
+(cl-defmethod related-files-apply ((jumper symbol) place)
+"Call JUMPER on PLACE."
   (funcall jumper place))
 
 (cl-defgeneric related-files-get-filler (jumper)
   "Return a filler associated with JUMPER."
   (get jumper 'related-files-filler))
 
+(cl-defgeneric related-files-place-exists-p (place)
+  "Return non-nil if PLACE exists.
+
+The default implementation assumes PLACE is a filename."
+  (file-exists-p place))
+
+(cl-defmethod related-files-place-exists-p ((place buffer))
+  "Call `buffer-live-p'."
+  (buffer-live-p place))
+
+(cl-defgeneric related-files-goto-place (place)
+  "Go to an existing PLACE.
+
+The default implementation assumes PLACE is a filename."
+  (find-file place))
+
+(cl-defgeneric related-files-goto-place ((place buffer))
+  "Call `switch-to-buffer'."
+  (switch-to-buffer place))
+
+(cl-defgeneric related-files-attach-jumper-to-place (jumper place)
```

Comment:
why do you need a method for this? Do you plan to override it?

Reply:
> Do you plan to override it?

Yes. because different places will need to have extra information (the jumper) 'attached' to them in different ways -- only filenames (strings) can be propertized (this was what we using before). The implementation in my init file of my org-roam jumper returns an org roam node as a place. This is a cl struct object, so can't be propertized, so I needed to write a different way of attaching something to it.

(I haven't got there yet, but soon I'll also write a corresponding method to *retrieve* the jumper *from* the place. I haven't got to generlising making new places yet though.)